File: app/src/main/java/com/tk4218/grocerylistr/Model/CalendarRecipes.java
package com.tk4218.grocerylistr.Model;

import com.tk4218.grocerylistr.Database.JSONResult;
import com.tk4218.grocerylistr.Database.QueryBuilder;
import com.tk4218.grocerylistr.Recipe;

import java.util.ArrayList;
import java.util.Date;

/*
 * Created by Tk4218 on 10/11/2017.
 */

public class CalendarRecipes {
    private QueryBuilder mQb = new QueryBuilder();

    private String mUsername;
    private Date mCalendarDate;
    private ArrayList<Recipe> mRecipes;

    public CalendarRecipes(String username, Date calendarDate){
        setUsername(username);
        setCalendarDate(calendarDate);

        JSONResult calendarRecipes = mQb.getCalendarRecipes(mUsername, calendarDate);
        setCalendarRecipes(calendarRecipes);
    }

    public CalendarRecipes(String username, Date calendarDate, ArrayList<Recipe> recipes){
        setUsername(username);
        setCalendarDate(calendarDate);
        setCalendarRecipes(recipes);
    }

    private void setUsername(String username){ mUsername = username; }

    public void setCalendarDate(Date calendarDate){
        mCalendarDate = calendarDate;
    }
    public Date getCalendarDate(){
        return mCalendarDate;
    }

    public void setCalendarRecipes(JSONResult calendarRecipes){
        mRecipes = new ArrayList<>();
        if(calendarRecipes.getCount() == 0) { return; }

        JSONResult breakfastRecipes = calendarRecipes.filter("MealType", "Breakfast");
        breakfastRecipes.sort("Sequence", JSONResult.SORT_ASCENDING);
        convertMeals(breakfastRecipes);

        JSONResult lunchRecipes = calendarRecipes.filter("MealType", "Lunch");
        breakfastRecipes.sort("Sequence", JSONResult.SORT_ASCENDING);
        convertMeals(lunchRecipes);

        JSONResult dinnerRecipes = calendarRecipes.filter("MealType", "Dinner");
        breakfastRecipes.sort("Sequence", JSONResult.SORT_ASCENDING);
        convertMeals(dinnerRecipes);
    }

    public void setCalendarRecipes(ArrayList<Recipe> calendarRecipes){
        mRecipes = calendarRecipes;
    }

    public void addRecipe(String recipeKey, String recipeName, String mealType){
        mRecipes.add(new Recipe(recipeKey, "", recipeName, mealType, "", "", false, 0, new Date(0), false));
    }

    public ArrayList<Recipe> getCalendarRecipes(){
        return mRecipes;
    }

    public Recipe getRecipe(String mealType, int sequence){
        int seq = 0;
        for (int i = 0; i < mRecipes.size(); i++) {
            if (mRecipes.get(i).getMealType().equals(mealType)){
                if(seq == sequence){
                    return mRecipes.get(i);
                }
                seq++;
            }
        }
        return null;
    }

    public ArrayList<Recipe> getRecipesByMealType(String mealType) {
        ArrayList<Recipe> recipes = new ArrayList<>();
        for (int i = 0; i < mRecipes.size(); i++) {
            if (mRecipes.get(i).getMealType().equals(mealType)) {
                recipes.add(mRecipes.get(i));
            }
        }
        return recipes;
    }

    private void convertMeals(JSONResult calendarRecipes){
        for(int i = 0; i < calendarRecipes.getCount(); i++) {
            addRecipe(calendarRecipes.getString(i, "RecipeKey"),
                      calendarRecipes.getString(i, "RecipeName"),
                      calendarRecipes.getString(i, "MealType"));
        }
    }
}

File: app/src/main/java/com/tk4218/grocerylistr/model/CalendarRecipes.kt
package com.tk4218.grocerylistr.model

import android.util.Log
import com.google.firebase.database.*
import com.tk4218.grocerylistr.fragments.CalendarLoadedCallback
import java.text.SimpleDateFormat
import java.util.*
import kotlin.collections.ArrayList
import kotlin.collections.HashMap

/*
 * Created by Tk4218 on 10/11/2017.
 */

data class CalendarRecipes(private var mUsername: String, var calendarDate: Date?, var recipes: ArrayList<Recipe>) {
    companion object {
        private var dateFormat = SimpleDateFormat("yyyy-MM-dd", Locale.US)

        fun addRecipe(userName: String, calendarDate: Date, recipeKey: String) {
            val database = FirebaseDatabase.getInstance()
            val recipeRef = FirebaseDatabase.getInstance().getReference("recipe/$recipeKey")

            recipeRef.addListenerForSingleValueEvent(object : ValueEventListener{
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    if(dataSnapshot.value == null) return

                    val recipe = dataSnapshot.getValue(Recipe::class.java)
                    val calendarRef = database.getReference("$userName/calendar/${dateFormat.format(calendarDate)}")

                    calendarRef.addListenerForSingleValueEvent(object : ValueEventListener {
                        override fun onDataChange(dataSnapshot: DataSnapshot) {
                            val type = object: GenericTypeIndicator<HashMap<String, Recipe>>() {}
                            var recipes = dataSnapshot.getValue(type)
                            if(recipes == null) recipes = HashMap()
                            recipes[recipe!!.recipeKey] = recipe
                            calendarRef.setValue(recipes)
                        }
                        override fun onCancelled(databaseError: DatabaseError) {
                            Log.e("CalendarRecipes", "Unable to Retrieve Calendar Recipes for $userName, ${dateFormat.format(calendarDate)}")
                        }
                    })
                }
                override fun onCancelled(databaseError: DatabaseError) {
                    Log.e("CalendarRecipes", "Unable to Retrieve Calendar Recipes for $userName, ${dateFormat.format(calendarDate)}")
                }
            })
        }

        fun removeRecipe(userName: String, calendarDate: Date, recipeKey: String) {
            val calendarRef = FirebaseDatabase.getInstance().getReference(userName).child("calendar").child(dateFormat.format(calendarDate)).child(recipeKey)
            calendarRef.addListenerForSingleValueEvent(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    if(dataSnapshot.value != null)
                        dataSnapshot.ref.removeValue()
                }
                override fun onCancelled(p0: DatabaseError) {
                    TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
                }
            })
        }

        fun getCalendar(userName: String, beginDate: Date, endDate: Date, calendarLoadedCallback: CalendarLoadedCallback){
            val startDate = clearDateTime(beginDate)
            val stopDate = clearDateTime(endDate)
            val calendarMap: HashMap<Date, CalendarRecipes> = HashMap()
            val date = Calendar.getInstance()
            date.time = startDate
            while(date.time < stopDate) {
                calendarMap[date.time] = CalendarRecipes(userName, date.time, ArrayList())
                date.add(Calendar.DAY_OF_MONTH, 1)
            }

            val calendarRef = FirebaseDatabase.getInstance().getReference("$userName/calendar")
            val query = calendarRef.orderByKey()
                            .startAt(dateFormat.format(startDate))
                            .endAt(dateFormat.format(stopDate))
            query.addListenerForSingleValueEvent(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    if(dataSnapshot.value != null){
                        val type = object: GenericTypeIndicator<HashMap<String, Recipe>>() {}
                        for(child in dataSnapshot.children)
                            calendarMap[dateFormat.parse(child.key)]!!.recipes = ArrayList(child.getValue(type)!!.values)
                    }
                    val calendarRecipes: ArrayList<CalendarRecipes> = ArrayList()
                    for(item in calendarMap.toSortedMap().values)
                        calendarRecipes.add(item)
                    calendarLoadedCallback.onCalendarLoaded(calendarRecipes)
                }
                override fun onCancelled(p0: DatabaseError) {
                }
            })
        }
        private fun clearDateTime(date: Date): Date {
            val calendar = Calendar.getInstance()
            calendar.time = date
            calendar.set(calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH), calendar.get(Calendar.DAY_OF_MONTH), 0, 0, 0)
            calendar.set(Calendar.MILLISECOND, 0)
            return calendar.time
        }
    }
}


